### PR TITLE
SpannerIO: parameterizing partitionQuery timeout

### LINKED ISSUE #25025: [Feature Request]: Increase Spanner PartitionQuery timeout to 5 minutes
### What would you like to happen?

The default 30s deadline is too small and may cause deadline exceeded errors when multiple PartitionQuery calls are running in parallel. This request is similar to [BEAM-11543](https://issues.apache.org/jira/browse/BEAM-11543). 

### Issue Priority

Priority: 2 (default / most feature requests should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [X] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by darshan-sj**
SpannerIO: parameterizing partitionQuery and partitionRead timeouts. fixes #25025 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by darshan-sj** [comment]
@pabloem Please review

**Comment by darshan-sj** [comment]
@nielm Please review

**Comment by nielm** [comment]
Changes LGTM. 

Suggestion: set the new timeouts by default to a (sensible) larger value in SpannerConfig, so that for most users, it will 'just work(TM)' without them needing to choose a value.

I am not sure what a sensible larger value should be - maybe check with @rbrtwng

**Comment by rbrtwng** [comment]
Thanks @nielm for the review. @darshan-sj and I have sync'ed and believed that not having a default deadline in SpannerIO makes most sense for now. That's, unless the user explicitly supplies a timeout, the default value from client library should be honored.  Here's the rationale: 

- Though setting a default to a sensible larger value in SpannerConfig will make it "just work", it might also hide some potential issues. 
-  We discussed with a few other engineers in Spanner team and generally we believe the current default deadline of 30s is too small - but PartitionQuery is mainly used in batch workload so it actually makes more sense to bump up the default deadline from the client library side, not from SpannerIO.  I will follow up separately on that. 

In short, I would support the current fix plan and give users the flexibility to "fix it" when they start seeing deadline exceeded errors.

**Comment by nielm** [comment]
Ok, makes sense LGTM.

**Comment by pabloem** [comment]
we no longer use ValuProvider parameters for transforms. These werre used for 'old' template formats, but they are no longer necessary for 'Flex' templates which is the main kind of template. Could you please remove the ValueProvider parameters and only use the direct value?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerConfig.java [hunk 1]
```diff
@@ -265,4 +273,24 @@ public SpannerConfig withRpcPriority(ValueProvider<RpcPriority> rpcPriority) {
...
+
+  /** Specifies the PartitionQuery timeout. */
+  public SpannerConfig withPartitionQueryTimeout(Duration partitionQueryTimeout) {
+    return withPartitionQueryTimeout(ValueProvider.StaticValueProvider.of(partitionQueryTimeout));
+  }
+
+  /** Specifies the PartitionQuery timeout. */
+  public SpannerConfig withPartitionQueryTimeout(ValueProvider<Duration> partitionQueryTimeout) {
+    return toBuilder().setPartitionQueryTimeout(partitionQueryTimeout).build();
+  }
```
Comment (pabloem) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerConfig.java:285]: avoid this. Just pass the Duration parameter directly. (ditto for others)

**Comment by pabloem** [comment]
Discussed offline. There are non-flex Spanner templates that the team wants to support.

LGTM
